torch/csrc/jit/serialization/export_module.cpp

```diff
@@ -522,21 +575,55 @@ class ScriptModuleSerializer {
       // If exception ocurrs we have a model with debug map that can be
       // used to symbolicate debug handles
       writeArchive("mobile_debug_handles", debug_info_telements);
+      static constexpr size_t kMinToCompress = 200;
+      // For delegated backends get source ranges that are in the debug info
+      // map. Since delegated backend replace original module with lowered
+      // module we will not serialize original module's code which is what would
+      // have contained source range. Since we dont have that anymore, extract
+      // source ranges out of delegated module and store in a separate archive.
+      // Note that we must do this first because in order to serialize inlined
+      // CS appropriate source_range_tags must have been generated.
+      auto backend_source_range_records = getBackendSourceRanges(module);
+      SourceRangePickler source_range_pickler;
+      updateSourceRangeTags(backend_source_range_records);
+      auto range_data = source_range_pickler.pickle(
+          backend_source_range_records, source_range_tags_);
+      std::string debugFilename = "delegated_backends.debug_pkl";
+      writer_.writeRecord(
+          debugFilename,
+          range_data.data(),
+          range_data.size(),
+          range_data.size() > kMinToCompress /*compress*/);
+
+      // For delegated backends get debug_info_map
+      // This is merged with other debug_info_map of other modules
+      // which were not delegated.
+      DelegateDebugInfoMapType backend_debug_info_map;
+      getBackendDebugInfoMap(module, backend_debug_info_map);
       // Now get the debug-handles-to-inlined-cs-ptr-map
       // And serialize that in a separate archive
       auto debug_handle_cs_ptr_map = debug_handle_manager.getCallStackPtrMap();
+      debug_handle_cs_ptr_map.insert(
+          backend_debug_info_map.begin(), backend_debug_info_map.end());
       CallStackDebugInfoPickler cs_debug_info_pickler;
       auto cs_data = cs_debug_info_pickler.pickle(
           debug_handle_cs_ptr_map, source_range_tags_);
       // Write out map: [debug-handle, {source range, InlinedCallStack}]
       std::string filename = "callstack_debug_map.pkl";
-      static constexpr size_t kMinToCompress = 200;
       writer_.writeRecord(
           filename,
           cs_data.data(),
           cs_data.size(),
           cs_data.size() > kMinToCompress /*compress*/);
     }
+    // This introduces a strange behavior where if you do
+    // module._save_for_mobile(..., debug_info=true)
+    // module._save_for_mobile(..., debug_info=true)
+    // Meaning if you try to save the model twice the second time
+    // you would lose the information.
+    // This should probably be fixed by having cleanup step perhaps
+    // inside the module dtor
```

Comment:
@SplitInfinity @raziel so something I realized is that doing cleanup of debug info like this leads to strange behavior.
For example:
```
Module model;
... do some lowering on some submodule...
module._save_for_mobile(.., with debug info);
// Now if you try to save it again with
module._save_for_mobile(.., with debug info);
// it would not work
```
Problem is that we do cleanup here and after that we have lost the lowered module's debug info. I wanted to get some opinions to see if this is ok behavior, my feeling is not, and what we can do.

In terms of what we can do:
If we store debug map in `class Module` then ownership of the debug map can be transferred to the lowered module class instead of us managing it. It does feel a bit intrusive of a change but to me that seems natural way of doing it.

Implication of this change on exception symbolication for the backends like glow where preprocess and execution happen in the same session:
Here in JIT interpreter we dont have access to the module that owns the graph as far as I know. One option, that I dont like too much, is that introduce a client side API, like `withLoweredModuleDebug` or some such, that consolidates all the lowered module's debug info from the corresponding `Module` instances and then debug handle from `DelegatedBackendException` can be symbolicated using this consolidated map. Map consolidation may not actually be copy of the map but just references to the map.
Any other option that you guys can think of?

Reply:
> If we store debug map in `class Module` then ownership of the debug map can be transferred to the lowered module class instead of us managing it. It does feel a bit intrusive of a change but to me that seems natural way of doing it.

`Module` is supposed to provide a C++ API for interacting with scripted `torch.nn.Modules`, so I don't think it makes sense for that debug info to live in that class. Since this debug info is exclusive to lowered modules, perhaps we could subclass `Module` (e.g. `LowereModule`) and put it there? Not sure if all uses of Module use reference/pointer  in order to automatically take advantage of polymorphism/inheritance though...

Reply:
I think that is a good idea. Probably lesser of the two evils. In this case my concern would be to have separate type introduced just for lowered module, that atm stores nothing but debug map. But it is better than what I was thinking.

> Not sure if all uses of Module use reference/pointer in order to automatically take advantage of polymorphism/inheritance though

@SplitInfinity did not follow this part.

Reply:
If there is user code the uses `Module` instead of `Module*` or `Module&`, then I think this hypothetical `LoweredModule` class would be sliced if assigned to a variable like `Module a = ...` and the debug info map and other extra state we add would not be available.

Reply:
What's the current lifecycle of Module (and consequently of having a LowereModule)? Basically when does it get destroyed?

But yes, high level it sounds reasonable that if LowereModule has extra data that needs to be preserved for serialization then having an actual class with the data embedded seems the natural solution.

In such case then it should then take the instance of BackendModuleDebugInfoRecorder (per my comments above). Then cleanup can happen in that class's destructor.

Reply:
I have commented elswhere but we can store `BackendDebugInfoMap` inside `LoweredModule` and that should take care of the lifetime.